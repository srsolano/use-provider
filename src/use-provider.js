import React, { createContext, useContext, useReducer } from 'react';
import PropTypes from 'prop-types';

const contextMap = new Map();

export function createProvider(initialState = null, reducer = state => state) {
  const StateContext = createContext(null);
  const DispatchContext = createContext(null);

  const Provider = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState);

    return (
      <DispatchContext.Provider value={dispatch}>
        <StateContext.Provider value={state}>{children}</StateContext.Provider>
      </DispatchContext.Provider>
    );
  };
  Provider.propTypes = { children: PropTypes.node };

  contextMap.set(Provider, { StateContext, DispatchContext });

  return Provider;
}

export function combineProviders(providers) {
  const [first, ...rest] = providers;
  return function Combined({ children }) {
    return rest.reduce((Combined, Provider) => {
      return (
        <Combined>
          <Provider>{children}</Provider>
        </Combined>
      );
    }, first);
  };
}

export function useProvider(Provider) {
  const { StateContext = {}, DispatchContext = {} } =
    contextMap.get(Provider) || {};
  return [useContext(StateContext), useContext(DispatchContext)];
}
