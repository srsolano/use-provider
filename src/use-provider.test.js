import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import { createProvider, useProvider, combineProviders } from './use-provider';

const reducer = (state, action) => {
  switch (action.type) {
    case 'update': {
      return action.payload;
    }

    default: {
      return state;
    }
  }
};

const createTestComponent = TestProvider => {
  return function TestComponent() {
    const [state, dispatch] = useProvider(TestProvider);
    return (
      <div>
        State: {`${state}`}
        <button onClick={() => dispatch({ type: 'update', payload: 'bar' })}>
          click
        </button>
      </div>
    );
  };
};

describe('createProvider', () => {
  afterEach(cleanup);

  it('should create a valid context provider', () => {
    const TestProvider = createProvider('foo', reducer);
    const TestComponent = createTestComponent(TestProvider);

    const { container } = render(
      <TestProvider>
        <TestComponent />
      </TestProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should create a default context provider', () => {
    const TestProvider = createProvider();
    const TestComponent = createTestComponent(TestProvider);

    const { container } = render(
      <TestProvider>
        <TestComponent />
      </TestProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should provide default state and dispatch values', () => {
    const TestProvider = createProvider();
    const { result } = renderHook(() => useProvider(TestProvider), {
      wrapper: TestProvider
    });

    const dispatch = result.current[1];
    act(() => {
      dispatch();
    });

    const state = result.current[0];
    expect(state).toBe(null);
  });
});

describe('combineProviders', () => {
  it('should create a combined provider', () => {
    const ProviderA = createProvider('a', reducer);
    const ProviderB = createProvider('b', reducer);
    const CombinedProvider = combineProviders([ProviderA, ProviderB]);
    const ComponentA = createTestComponent(ProviderA);
    const ComponentB = createTestComponent(ProviderB);

    const { container } = render(
      <CombinedProvider>
        <ComponentA />
        <ComponentB />
      </CombinedProvider>
    );

    expect(container).toMatchSnapshot();
  });
});

describe('useProvider', () => {
  it('should return context state', () => {
    const TestProvider = createProvider('foo', reducer);
    const { result } = renderHook(() => useProvider(TestProvider), {
      wrapper: TestProvider
    });
    const state = result.current[0];
    expect(state).toBe('foo');
  });

  it('should return context dispatch', () => {
    const TestProvider = createProvider('foo', reducer);
    const { result } = renderHook(() => useProvider(TestProvider), {
      wrapper: TestProvider
    });

    const dispatch = result.current[1];
    act(() => {
      dispatch({ type: 'update', payload: 'bar' });
    });

    const state = result.current[0];
    expect(state).toBe('bar');
  });

  it('should not crash if is called with no params', () => {
    const { result } = renderHook(() => useProvider());
    expect(result.current).toEqual([undefined, undefined]);
  });
});
