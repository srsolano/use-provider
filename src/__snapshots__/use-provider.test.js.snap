// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combineProviders should create a combined provider 1`] = `
<div>
  <div>
    State: 
    a
    <button>
      click
    </button>
  </div>
  <div>
    State: 
    b
    <button>
      click
    </button>
  </div>
</div>
`;

exports[`createProvider should create a default context provider 1`] = `
<div>
  <div>
    State: 
    null
    <button>
      click
    </button>
  </div>
</div>
`;

exports[`createProvider should create a valid context provider 1`] = `
<div>
  <div>
    State: 
    foo
    <button>
      click
    </button>
  </div>
</div>
`;
